File: Haskell/HelloWorld/src/Kex2/Kex2Battle.hs
{-# LANGUAGE NamedFieldPuns #-}
module Kex2.Kex2Battle where

import Prelude
import Control.Monad ((>=>))
import Data.Text (Text, pack, unpack)
import Shared.Player (Player(..))
import Utility (getInt, getRandomInteger)
import Shared.Enemy (makeEnemy, EnemyType(..), Enemy(..))
import Shared.Constants (attackHit, skillHit, magicHit)

data AttackMethod
  = Attack Integer Integer
  | Skill Integer Integer
  | Magic Integer Integer

data ProgressResult
  = End Text
  | Continue Player

gameLoop :: (Show a, Num a) => Player -> a -> Int -> IO Text
gameLoop Player{ playerHp } kills choice | playerHp <= 0 || choice == 0 = pure $ pack $ "リ〇ミト！\n戦闘回数：" <> show kills <> "回　残りHP：" <> show playerHp
gameLoop player@Player{ playerHp } kills _ = do
  putStrLn $ "\n現HP：" <> show playerHp
  putStrLn "奥に進みますか？（１：奥に進む　０．帰る）＞"
  choice <- getInt
  case choice of
    0 -> gameLoop player kills choice
    _ -> do
      ordinal <- getRandomInteger 2
      enemy <- makeEnemy ordinal
      progressResult <- engageBattle (extractEnemyData enemy) player
      case progressResult of
        End txt -> do
          putStrLn $ unpack txt
          gameLoop player{ playerHp = 0 } kills 0
        Continue pl -> gameLoop pl (kills + 1) 1

extractEnemyData :: EnemyType -> Enemy
extractEnemyData enemy = case enemy of
  Golem e -> e
  Goblin e -> e
  Slime e -> e

engageBattle :: Enemy -> Player -> IO ProgressResult
engageBattle enemy@Enemy{ enemyName, enemyLevel } player = do
  putStrLn $ unpack enemyName <> "Lv." <> show (enemyLevel + 1) <> "が現れた！"
  battleLoop enemy player

battleLoop :: Enemy -> Player -> IO ProgressResult
battleLoop Enemy{ enemyName, enemyLevel, enemyHp } player | enemyHp <= 0 = do
  putStrLn $ unpack enemyName <> "Lv." <> show (enemyLevel + 1) <> "を倒した！"
  pure $ Continue player
battleLoop Enemy{ enemyName } Player{ playerHp } | playerHp <= 0 = do
  putStrLn $ "あなたは" <> unpack enemyName <> "に負けました！"
  pure $ End $ pack "ゲームオーバー！"
battleLoop enemy@Enemy{ enemyName, enemyHp } player = do
  putStrLn $ unpack enemyName <> " 残りHP：" <> show enemyHp
  putStrLn "武器を選択してください（１．攻撃　２．特技　３．魔法）＞"
  attackMethod <- getInt >>= \i -> selectAttack i
  newEnemy <- damageEnemy attackMethod enemy
  putStrLn $ unpack enemyName <> "の攻撃！"
  newPlayer@Player{ playerHp = newPlayerHp } <- damagePlayer enemy player
  if newPlayerHp > 0 then do
    putStrLn $ "プレイヤー残りHP：" <> show newPlayerHp
    battleLoop newEnemy newPlayer
  else
    battleLoop newEnemy newPlayer

checkHitOrMiss :: Integer -> IO Bool
checkHitOrMiss hit = getRandomInteger 100 >>= \i -> pure $ i < hit

selectAttack :: (Eq t, Num t) => t -> IO AttackMethod
selectAttack 1 = getRandomInteger 40 >>= \i -> pure $ Attack (60 + i) attackHit
selectAttack 2 = getRandomInteger 100 >>= \i -> pure $ Skill (30 + i) skillHit 
selectAttack 3 = getRandomInteger 180 >>= \i -> pure $ Magic (20 + i) magicHit 
selectAttack _ = selectAttack 1

damageEnemy :: AttackMethod -> Enemy -> IO Enemy
damageEnemy attackMethod enemy@Enemy{ enemyHp, enemyFlee, enemyDefense } = do
  let (dmg, hit) = case attackMethod of
        Attack d h -> (d, h)
        Skill d h -> (d, h)
        Magic d h -> (d, h)
  hitOrMiss <- checkHitOrMiss $ hit - enemyFlee
  if hitOrMiss then do
    let actualDamage = if dmg - enemyDefense < 0 then 0 else dmg - enemyDefense
    putStrLn $ show actualDamage <> "のダメージ！"
    let newEnemyHp = if enemyHp - actualDamage < 0 then 0 else enemyHp - actualDamage
    pure $ enemy{ enemyHp = newEnemyHp }
  else do
    putStrLn "攻撃を外した！"
    pure enemy

damagePlayer :: Enemy -> Player -> IO Player
damagePlayer Enemy{ enemyHit, enemyAttack } player@Player{ playerHp, playerDefense } = do
  hitOrMiss <- checkHitOrMiss enemyHit
  if hitOrMiss then do
    let injury = if enemyAttack - playerDefense < 0 then 0 else enemyAttack - playerDefense
    putStrLn $ show injury <> "のダメージ！"
    let newPlayerHp = if playerHp - injury < 0 then 0 else playerHp - injury
    pure player{ playerHp = newPlayerHp }
  else do
    putStrLn "攻撃を外した！"
    pure player

File: FSharp/HelloWorld/Lib/Kex2/Kex2Battle.fs
module HelloWorld.Lib.Kex2.Kex2Battle

open System
open HelloWorld.Lib.Shared
open HelloWorld.Lib.Shared.Enemy
open Microsoft.FSharp.Core

type AttackMethod =
    Attack of (int * int)
    | Skill of (int * int)
    | Magic of (int * int)
    
type ProgressResult = Continue of Player | End of string

let rec private SelectAttack choice =
    let rng = Random()
    match choice with
    | 1 -> Attack (rng.Next(0, 41), Constants.AttackHit)
    | 2 -> Skill (rng.Next(0, 101), Constants.SkillHit)
    | 3 -> Magic (rng.Next(0, 181), Constants.MagicHit)
    | _ -> SelectAttack 1
    
let private CheckHitOrMiss hit = Random().Next(0, 101) < hit

let private ExtractEnemyData = function
    | Golem e -> e
    | Goblin e -> e
    | Slime e -> e

let private DamageEnemy attackMethod (enemy: Enemy) =
    let dmg, hit = match attackMethod with
                        | Attack item -> item
                        | Skill item -> item
                        | Magic item -> item
    let hitOrMiss = CheckHitOrMiss hit
    if hitOrMiss then
        let actualDamage = if dmg - enemy.Defense < 0 then 0 else dmg - enemy.Defense
        printfn $"{actualDamage}のダメージ！"
        let newEnemyHp = if enemy.Hp - actualDamage < 0 then 0 else enemy.Hp - actualDamage
        enemy.Hp <- newEnemyHp
        enemy
    else
        printfn "攻撃を外した！"
        enemy
        
let private DamagePlayer enemy (player: Player) =
    let hitOrMiss = CheckHitOrMiss enemy.Hit
    if hitOrMiss then
        let injury = if enemy.Attack - player.Defense < 0 then 0 else enemy.Attack - player.Defense
        printfn $"{injury}のダメージ！"
        let newPlayerHp = if player.Hp - injury < 0 then 0 else player.Hp - injury
        player.Hp <- newPlayerHp
        player
    else
        printfn "攻撃を外した！"
        player
        
let rec private BattleLoop enemy player =
    if enemy.Hp <= 0 then
        printfn $"{enemy.Name}Lv.${enemy.Level + 1}を倒した！"
        Continue player
    else
        match player.Hp with
        | x when x <= 0 ->
            printfn $"あなたは{enemy.Name}に負けました！"
            End "ゲームオーバー！"
        | _ ->
            printfn $"{enemy.Name} 残りHP：{enemy.Hp}"
            printf "武器を選択してください（１．攻撃　２．特技　３．魔法）＞"
            let choice = ref 0
            let userInput = Console.ReadLine()
            let parseResult = Int32.TryParse(userInput, choice)
            if not parseResult then
                BattleLoop enemy player
            else
                let attackMethod = SelectAttack choice.Value
                let newEnemy = DamageEnemy attackMethod enemy
                printfn $"{enemy.Name}の攻撃！"
                let newPlayer = DamagePlayer enemy player
                if newPlayer.Hp > 0 then
                    printfn $"プレイヤー残りHP：{newPlayer.Hp}"
                    BattleLoop newEnemy newPlayer
                else
                    BattleLoop newEnemy newPlayer
                    
let private EngageBattle enemy player =
    printfn $"{enemy.Name}Lv.{enemy.Level}が現れた！"
    BattleLoop enemy player
    
let GameLoop player =
    let rec innerLoop (player': Player) kills = function
        | 0 -> $"リ〇ミト！\n戦闘回数：{kills}回　残りHP：{player'.Hp}"
        | _ when player'.Hp <= 0 -> innerLoop player' kills 0
        | _ ->
            printfn $"\n現HP：{player'.Hp}"
            printf "奥に進みますか？（１：奥に進む　０．帰る）＞"
            let choice = ref 0
            let userInput = Console.ReadLine()
            let parseResult = Int32.TryParse(userInput, choice)
            if not parseResult then
                innerLoop player' kills 1
            else
                match choice.Value with
                | 0 -> innerLoop player' kills choice.Value
                | _ ->
                    let ordinal = Random().Next(0, 3)
                    let enemy = MakeEnemy ordinal
                    match EngageBattle (ExtractEnemyData enemy) player' with
                    | End s ->
                        Console.WriteLine s
                        player'.Hp <- 0
                        innerLoop player' kills 0
                    | Continue p -> innerLoop p (kills + 1) 1
    innerLoop player 0 1